File: go/v2/crypto/hkdf.go
package crypto

import (
	"crypto/sha256"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
)

// HKDFDerive 计算 HKDF-SHA256(IKM, salt, info, length)。
//
// 当 salt 为空时，按 RFC 5869 §2.2 默认填充为 HashLen 个零字节（SHA-256 即 32 字节）。
// 这与 Python `cryptography` HKDF 在 salt=None 时的行为一致。
func HKDFDerive(ikm, salt, info []byte, length int) ([]byte, error) {
	if length <= 0 {
		return nil, fmt.Errorf("HKDF: length 必须为正，实际 %d", length)
	}
	if len(salt) == 0 {
		salt = make([]byte, sha256.Size)
	}
	r := hkdf.New(sha256.New, ikm, salt, info)
	out := make([]byte, length)
	if _, err := io.ReadFull(r, out); err != nil {
		return nil, fmt.Errorf("HKDF: 读取派生输出失败: %w", err)
	}
	return out, nil
}

File: js/src/v2/crypto/hkdf.ts
/**
 * AUN E2EE V2: HKDF-SHA256
 *
 * 规范引用: §3.2 / §5.2 / §10
 *
 * 浏览器实现：使用 WebCrypto subtle.deriveBits 完成 HKDF。
 * - salt 长度为 0 时按 RFC 5869 规则替换为 32 字节零（与 Python sdk 行为一致）。
 * - info 必须是 BufferSource（Uint8Array 即可）。
 * - 输出长度由 `length` 控制（字节数）。
 */

/**
 * 计算 HKDF-SHA256(ikm, salt, info, length)。
 *
 * @param ikm    输入密钥材料
 * @param salt   盐（可空）
 * @param info   上下文/信息字段
 * @param length 输出字节数
 * @returns      派生密钥
 */
export async function hkdfSha256(
  ikm: Uint8Array,
  salt: Uint8Array,
  info: Uint8Array,
  length: number,
): Promise<Uint8Array> {
  const realSalt = salt.length === 0 ? new Uint8Array(32) : salt;
  // ikm 必须是独立 ArrayBuffer，避免视图歧义
  const baseKey = await crypto.subtle.importKey(
    'raw',
    ikm.slice().buffer,
    'HKDF',
    false,
    ['deriveBits'],
  );
  const bits = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: realSalt.slice().buffer,
      info: info.slice().buffer,
    } as HkdfParams,
    baseKey,
    length * 8,
  );
  return new Uint8Array(bits);
}
